test/distributed/test_c10d_nccl.py

```diff
@@ -2593,6 +2593,100 @@ def test_nccl_timeout(self):
+        extern "C" {
+          // Note that windows needs __declspec(dllexport): https://stackoverflow.com/a/24575865
+          C10_EXPORT void* nccl_alloc(size_t size, int device, void* stream) { 
+            std::cout << "Using ncclMemAlloc" << std::endl;
+            void* ptr;
+            ncclResult_t err = ncclMemAlloc(&ptr, size);
+            return ptr;
+          }
+          C10_EXPORT void nccl_free(void* ptr, size_t size, int device, void* stream) {
+            std::cout << "Using ncclMemFree" << std::endl;
+            ncclResult_t err = ncclMemFree(ptr);
+          }
+        }
+        """
+        nccl_allocator_libname = "nccl_allocator"
+        nccl_allocator = load_inline(
+            name=nccl_allocator_libname,
+            cpp_sources=nccl_allocator_source,
+            with_cuda=True,
+            extra_ldflags=["-lnccl"],
+            is_python_module=False,
+            keep_intermediates=False,
+            verbose=True,
+        )
+        return nccl_allocator
+
+    def setUp(self):
+        super().setUp()
+        # TORCH_NCCL_BLOCKING_WAIT overrides TORCH_NCCL_ASYNC_ERROR_HANDLING hence tests
+        # that use TORCH_NCCL_BLOCKING_WAIT will test it as expected.
+        os.environ["TORCH_NCCL_ASYNC_ERROR_HANDLING"] = "1"
+        nccl_debug_file = tempfile.NamedTemporaryFile()
+        os.environ["NCCL_ALGO"] = "NVLS"
+        os.environ["NCCL_DEBUG"] = "INFO"
+        os.environ["NCCL_DEBUG_SUBSYS"] = "NVLS"
+        os.environ["NCCL_DEBUG_FILE"] = nccl_debug_file.name
+        os.environ["ALLOCATOR_PATH"] = self.createNcclAllocator()
+        self._spawn_processes()
+
+    def tearDown(self):
+        super().tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    @requires_nccl()
+    @requires_nccl_version(
+        (2, 19), "Need NCCL 2.19 for user buffer registration"
+    )
+    @skip_if_lt_x_gpu(4)
+    def test_nccl_user_buffer_registration(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        c10d.init_process_group(
+            backend="nccl", rank=self.rank, world_size=self.world_size, store=store
+        )
+        device = torch.device(f"cuda:{self.rank}")
+        torch.cuda.set_device(self.rank)
+        pg = c10d.distributed_c10d._get_default_group()
+        backend = pg._get_backend(torch.device(device))
+        allocator = torch.cuda.memory.CUDAPluggableAllocator(
+            os.environ["ALLOCATOR_PATH"],
+            "nccl_alloc",
+            "nccl_free",
+        )
+        pool = torch.cuda.MemPool(allocator.allocator())
+
+        # register buffers and use NVLS reductions
+        with torch.cuda.use_mem_pool(pool):
+            tensor = torch.arange(1024 * 1024 * 2, device=device)
+            backend.register_user_buffers(device)
+            pg.allreduce(tensor).wait()
+            torch.cuda.synchronize(device=device)
+            backend.deregister_user_buffers(device)
```

Comment:
See my comment below re API name. Would something like `register_memory(device)` or `register_mem_pool(device)` be more accurate?

Reply:
And, is it possible to do:
```
pool = torch.cuda.MemPool(allocator.allocator())
backend.register_mem_pool(pool)

with torch.cuda.use_mem_pool(pool):
    ...

backend.deregister_mem_pool(pool)
pool.release()
pool.empty_cache()
```

Reply:
Now it is possible to do the above. Note that registration has to come after the memory has been allocated into the pool, because of how `registerMemPool` is written. `registerMemPool` finds all the blocks currently in a pool and then calls `ncclCommRegister` on them. So if it's called before any allocation, that means it's registering an empty pool.